done

bethanyhaubert opened the pull request.

bethanyhaubert committed:
done

bethanyhaubert committed:
added overall winner

bethanyhaubert committed:
added some puts

bethanyhaubert committed:
broke out methods

bethanyhaubert committed:
changed layout

bethanyhaubert committed:
added moar tests

bethanyhaubert committed:
added 'Hal'

bethanyhaubert committed:
added global variables

bethanyhaubert committed:
added lizard & spock

bethanyhaubert committed:
added version 02 file

bethanyhaubert committed:
two working games

bethanyhaubert committed:
added class

bethanyhaubert committed:
testing tests

bethanyhaubert commented:
My test is still giving me problems.
I moved the instance variables that were requiring git.chomp into their own method. 
1) Error:
RockPaperScissorsTest#test_player_one_win:
NoMethodError: undefined method `find_winner' for #<RockPaperScissorsTest:0x007f9f9b062150>
    /Users/beth/Code/ruby-app__rock-paper-scissors/tests/rps_v2_test.rb:12:in `test_player_one_win'

sumeetjain commented:
This is related to something we went over just now, actually. (That's not say you should just go watch the video. I'm just letting you know that it's there for additional information, if my response below isn't enough.)

The `find_winner` method is defined within the `RockPaperScissorsGame` class. It cannot be called except on a `RockPaperScissorsGame` object.

So if you just do `x = find_winner`, then you'll get an error. Like doing `x = upcase`.

You need to call the method on a `RockPaperScissorsGame` object.

bethanyhaubert commented:
I did.

```
	def test_player_one_win
		sample_game =  RockPaperScissorsGame.new
		sample_game.set_p1_weapon("rock")
		sample_game.set_p2_weapon("scissors")
		win_hash = {"scissors" => ["rock", "spock"], "rock" => ["paper", "spock"], "paper" => ["scissors", "lizard"] }

	  	assert_equal(@player1, find_winner(win_hash))
	end
```

sumeetjain commented:
I did, actually.

You're calling `find_winner` in your test. But you cannot call `find_winner` like that. It must be called upon a `RockPaperScissorsGame` object.

You _are_ creating a `RockPaperScissorsGame` object–you store it in `sample_game`. But you're not using `sample_game` at all in your assertion.

bethanyhaubert committed:
testing test again

bethanyhaubert commented:
I tried that. still having issues. Can you help troubleshoot?

sumeetjain commented:
Nice–you've fixed that, so now you have a new error. (That's good–new errors are progress.)

_Note: I would like you to practice asking questions with details like below–it's okay if it's not perfect, but do give it a try when you have errors._

The error now is stating `undefined method `[]' for nil:NilClass` and refers you to your `find_winner` method, where you are running `if @win_hash[@p1].include?(@p2)`.

So it's saying that you're trying to do `[]` on `nil`. That must mean that Ruby thinks `@win_hash` is `nil`. Because you're doing `@win_hash[@p1]`, which Ruby is interpreting as `nil[@p1]`.

So the question now is, "Why is `@win_hash` not set?"

Where are you defining `@win_hash`? Has that code run in your test yet? Because if it has not, then it makes sense that `@win_hash` would be `nil`.

bethanyhaubert commented:
I moved @win_hash outside of any methods. Now it's just located in the RockPaperScissorsGame class. But I'm getting the same error. I don't know what else to add/delete

   1) Error:
RockPaperScissorsTest#test_player_one_win:
NoMethodError: undefined method `[]' for nil:NilClass
    /Users/beth/Code/ruby-app__rock-paper-scissors/lib/rps_v2.rb:134:in `find_winner'
    /Users/beth/Code/ruby-app__rock-paper-scissors/tests/rps_v2_test.rb:11:in `test_player_one_win'

bethanyhaubert committed:
more test stuff

bethanyhaubert commented:
Despite fixing all of the issues you saw, when I run rake I'm getting this error: 
  1) Error:
RockPaperScissorsTest#test_player_one_win:
NoMethodError: undefined method `[]' for nil:NilClass
    /Users/beth/Code/ruby-app__rock-paper-scissors/lib/rps_v2.rb:133:in `find_winner'
    /Users/beth/Code/ruby-app__rock-paper-scissors/tests/rps_v2_test.rb:11:in `test_player_one_win'

it would be nice if you could help me figure out what the fuck is wrong with it because i don't know.

bethanyhaubert committed:
put win_hash in it's own method

sumeetjain commented:
> Despite fixing all of the issues you saw

Not for the sake of my ego, but more for the sake of your practice and growth, consider being more cautious with such assumptions. You think you've fixed all of the issues I saw. I disagree. It will be more productive if your perspective when you encounter a bug is _not_, "I've fixed everything." but rather, "What am I misunderstanding?"

As for your code:

You've defined a new method `set_win_hash` for setting the win_hash initially. That's great–you're absolutely right that you'll need to do that, as otherwise `@win_hash` is `nil` (and as I mentioned before that's a problem).

But where do you _call_ that method? When is it being run? I see it being run in the `game_on` method. That will work just fine when you're playing the game. But do your tests run that method? It doesn't look like it (and nor should they).

Your `test_player_one_win` method does various things to **set up a good sample game**. You're already doing some things there, `sample_game.set_p1_weapon("rock")`. Did you consider that you would need to set the win hash there as well?

bethanyhaubert commented:
Yes, i had the win_hash inside of the test previously

sumeetjain commented:
Show me that code please.

bethanyhaubert commented:
https://www.dropbox.com/s/k5ss87bbd8joe70/Screenshot%202016-02-01%2016.15.04.png?dl=0

bethanyhaubert committed:
added win_hash back into test

bethanyhaubert commented:
And i just tried it again.

sumeetjain commented:
That was a while ago. Since then, we've been over the need to have a method that sets `@win_hash`. **Where are you calling that method?**

bethanyhaubert commented:
It's a method inside RockPaperScissorsGame class. calling that method inside the game_on method

bethanyhaubert commented:
```
def set_win_hash
		@win_hash = {"paper" => ["scissors", "lizard"], "scissors" => ["rock", "spock"], "rock" => ["paper", "spock"], "lizard" => ["rock", "scissors"], "spock" => ["paper", "lizard"]}
	end

```

bethanyhaubert commented:
i fixed it

sumeetjain commented:
Right. And as I said before:

> I see it being run in the `game_on` method. That will work just fine when you're playing the game. But do your tests run that method? It doesn't look like it (and nor should they).

You've bought into the [correct] idea that a test needs to build up a sample piece of information. You're already doing that correctly:

```ruby
def test_player_one_win
  sample_game =  RockPaperScissorsGame.new
  sample_game.set_p1_weapon("rock")
  sample_game.set_p2_weapon("scissors")
  sample_game.set_win_hash

  assert_equal(@player1, sample_game.find_winner)
end
```

We then went over how your `sample_game` object needs to have its `@win_hash` set in order for the `find_winner` method to work.

The `@win_hash` is set in the `game_on` method, **but your tests do not run that method**. Because if they did, you'd have a bunch of `gets` prompts in your tests.

So you made a method called `set_win_hash`. You _do_ still use that in your `game_on` method. But, again, your `game_on` method is not of use in the test.

So you need to set the win hash in the test. Just like you're setting each player's weapon.

sumeetjain commented:
If you read this carefully, think about it, be humble about your own understanding of the code, and are still unable to get your test passing, come see me and we can patiently work together through it. I'm here for you.

bethanyhaubert commented:
i don't need set_game_hash in my game_on method. I only need it in my find_winner hash, so i called it inside that method only

sumeetjain commented:
I do not see the `set_game_hash` method being called inside your `find_winner` method.

If you would like to come up, I would welcome a chance to go over this more directly.

bethanyhaubert committed:
fixing last two tests. can't find problem

bethanyhaubert commented:
"I do not see the set_game_hash method being called inside your find_winner method."
i see it. i fixed that first test 20 minutes ago. see  "i fixed it" comment above.

bethanyhaubert commented:
i got the first three tests working. only the last one isn't working

bethanyhaubert commented:
it's the same problem i had last night, but i don't remember how i fixed it

bethanyhaubert commented:
```
	def test_overall_draw_false
		sample_game =  RockPaperScissorsGame.new
		@player1 = "Beth"
		@player2 = "Hal"
		score = {@player1 => 2, @player2 => 3}

		assert_equal(false, sample_game.overall_draw?(score))
	end
```

should be returning false, but it's not.

sumeetjain commented:
If you edit your code locally but don't push it to GitHub, I have no way of seeing what you see.

When I made my comment, you had not pushed your code.

![](http://cl.ly/f1Rv/Screen%20Shot%202016-02-01%20at%204.43.20%20PM.png)

Notice how my comment appears before your commit. **That** is the commit in which you added `set_game_hash` to the `find_winner` method. Glad you figured that out–nice work.

sumeetjain commented:
Re: Your latest code, instance variables cannot be set except by methods in the class. So if you want to set `@player1`, you have to use some method in the class to do that.

Same for accessing instance variables. They can't be accessed except by running some method on the object.

Moving to another student now, but feel free to ask clarification questions.

bethanyhaubert committed:
done

bethanyhaubert committed:
done

sumeetjain commented on tests/rps_v2_test.rb:
```diff
@@ -0,0 +1,50 @@
+		sample_game =  RockPaperScissorsGame.new
+		sample_game.set_p1_weapon("rock")
+		sample_game.set_p2_weapon("scissors")
+		@win_hash = {"paper" => ["scissors", "lizard"], "scissors" => ["rock", "spock"], "rock" => ["paper", "spock"], "lizard" => ["rock", "scissors"], "spock" => ["paper", "lizard"]}
```
You can't set an object's instance variables like this. If you want to set or change the value of `sample_game`'s `@win_hash` property, **the only** way to do it is using a method in the `RockPaperScissorsGame` class that does that. There are _no_ exceptions to this.

If you don't have a method in the `RockPaperScissorsGame` that sets `@win_hash`, then you should make one.

In your case, you _do_ have such a method. Great! :+1: You would use that method here instead of `@win_hash = {"paper" => ["sc...`. Because, and I need to stress this: You can't set an object's instance variables except by using one of that object's methods. It's _impossible_.

But now the question is, **why** are you setting `@win_hash` here at all? I know I suggested it, but you found an alternative approach, which is to simply call your `set_win_hash` method from inside the `find_winner` method. Smart. But again, why bother with `@win_hash = {"paper" => ["scissors", "liz...` then?

sumeetjain commented on tests/rps_v2_test.rb:
```diff
@@ -0,0 +1,50 @@
+		sample_game.set_p1_weapon("rock")
+		sample_game.set_p2_weapon("scissors")
+		@win_hash = {"paper" => ["scissors", "lizard"], "scissors" => ["rock", "spock"], "rock" => ["paper", "spock"], "lizard" => ["rock", "scissors"], "spock" => ["paper", "lizard"]}
+	  	assert_equal(@player1, sample_game.find_winner)
```
It is **impossible** to access an instance variable except through an object's method. If you had a method in your `RockPaperScissorsGame` class like this:

``` ruby
def get_player_1
  return @player1
end
```

Then you could run `sample_game.get_player_1`. And it would return the requested information.

But you should understand that it is **impossible** to simply ask, "What's `@player1`?" – it will just return `nil`.

So even if your tests are passing, you should investigate how/why. Because `@player1` is definitely `nil`.

bethanyhaubert committed:
removed @hash_index from tests

bethanyhaubert commented:
fixed it

bethanyhaubert committed:
not sure how to deal with instance variables now that i have two classes

bethanyhaubert commented:
I am using the instance variables @player1 and @player2 in my RPSGame class and I'm trying to assign names to those variables by calling Player class methods inside the RPSGame class method, but when I run my game_on method, the hash where i track my @player1 and @player2 scores are showing nil for the keys. Or if I'm playing against the computer, {nil=>0, "Hal"=>0}.

It seems like doing this:
```
		@player1 = Player.new
		@player1.set_p1_name(@player1.get_player_name)
```
And then this:
`		score = {@player1 => 0, @player2 => 0}
`
should be working so my score hash is storing the name from above in the hash, but it's not. The lines of code above are all from my game_on method.

sumeetjain commented:
What specifically is happening instead of it working?

That is, I understand your question (You laid it out very well–thank you). But I'd like to know what error you're getting, or what incorrect result you're getting, etc.

bethanyhaubert commented:
Names that are input are not being stored where I thought they were. When I enter my name, nothing is returned. See game play below:

```
If you would like to play against the computer, please type 'yes'.
yes
Player one, please type your name.

-------------------------------------------------------
beth
Please type how many rounds you would like to play.
1
, please type rock, paper, scissors, lizard, spock.

-------------------------------------------------------
rock
 chose rock. Hal chose lizard.
It's a draw!
{nil=>0, "Hal"=>0}
Player 1 score is 0
Player 2 score is 0
No one wins overall game.
```

sumeetjain commented on lib/rps_v2.rb:
```diff
@@ -0,0 +1,158 @@
+		score = {@player1 => 0, @player2 => 0}
+
+		while round_number < number_of_games
+			puts "#{@player1}" +", please type rock, paper, scissors, lizard, spock."
```
`@player1` is the entire Player object–not just the name. The name is stored inside the Player object though, so you can write a method in the Player class that returns the name property.

Let's say you name that method `name_value`. You'd then be able to change this line to `puts "#{@player1.name_value}"`.

sumeetjain commented:
I'm still not sure why exactly you're getting the `nil` values that you seem to be. I'm taking a closer look.

bethanyhaubert committed:
adding methods to new class and calling them in first class

bethanyhaubert committed:
class 2 testing
